File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Events/EventBus+ModelLifecycle.swift
//
//  EventBus+ModelLifecycle.swift
//  RunAnywhere SDK
//
//  Typed model-lifecycle stream over the raw SDK event bus.
//
//  Native commons signals "model loaded/unloaded" on THREE different proto
//  channels depending on the path that performed the work:
//    1. component-lifecycle events (EVENT_CATEGORY_COMPONENT,
//       `event.componentLifecycle.currentState`) — the `loadModel` path,
//    2. model events (`event.model.kind` load/unload completed),
//    3. LLM generation events (`event.generation.kind` modelLoaded/Unloaded).
//
//  Which channel fires when is an SDK-internal detail. Consumers should not
//  need to know it — before this helper every app ViewModel hand-decoded all
//  three with an identical switch. `EventBus.modelLifecycle` folds them into
//  one typed stream; `modelLoaded` / `modelUnloaded` are pre-filtered views.
//

import Combine

// MARK: - Typed change

/// One model load/unload transition, decoded from the raw event bus.
public struct RAModelLifecycleChange: Sendable {
    public enum Kind: Sendable {
        case loaded
        case unloaded
    }

    /// Whether the model finished loading or was unloaded.
    public let kind: Kind

    /// Registry id of the affected model. May be empty when the native
    /// channel did not carry one (rare; treat as "current model").
    public let modelID: String

    /// SDK component slot the change applies to (.llm, .stt, .tts, ...).
    public let component: RASDKComponent

    /// The underlying raw event, for consumers that need extra payload
    /// fields (progress, framework, error, ...).
    public let event: RASDKEvent
}

// MARK: - EventBus typed publishers

public extension EventBus {

    /// Unified model load/unload stream across all native signal channels.
    ///
    /// ```swift
    /// RunAnywhere.events.modelLifecycle
    ///     .filter { $0.component == .llm }
    ///     .receive(on: DispatchQueue.main)
    ///     .sink { change in ... }
    /// ```
    var modelLifecycle: AnyPublisher<RAModelLifecycleChange, Never> {
        events
            .compactMap(EventBus.modelLifecycleChange(from:))
            .eraseToAnyPublisher()
    }

    /// `modelLifecycle` filtered to load completions.
    var modelLoaded: AnyPublisher<RAModelLifecycleChange, Never> {
        modelLifecycle
            .filter { $0.kind == .loaded }
            .eraseToAnyPublisher()
    }

    /// `modelLifecycle` filtered to unloads.
    var modelUnloaded: AnyPublisher<RAModelLifecycleChange, Never> {
        modelLifecycle
            .filter { $0.kind == .unloaded }
            .eraseToAnyPublisher()
    }

    /// Decode a raw SDK event into a lifecycle change, or nil when the event
    /// is not a load/unload transition. Exposed so non-Combine consumers can
    /// reuse the exact same channel mapping.
    static func modelLifecycleChange(from event: RASDKEvent) -> RAModelLifecycleChange? {
        // Channel 1: component-lifecycle (the canonical loadModel path).
        if event.category == .component {
            let lifecycle = event.componentLifecycle
            switch lifecycle.currentState {
            case .ready:
                return RAModelLifecycleChange(
                    kind: .loaded,
                    modelID: lifecycle.modelID,
                    component: lifecycle.component,
                    event: event
                )
            case .notLoaded, .unloading, .shutdown, .deleting:
                return RAModelLifecycleChange(
                    kind: .unloaded,
                    modelID: lifecycle.modelID,
                    component: lifecycle.component,
                    event: event
                )
            default:
                return nil
            }
        }

        // Channels 2 + 3: model events and LLM generation events.
        let modelID = event.model.modelID.isEmpty
            ? event.generation.modelID
            : event.model.modelID

        switch (event.model.kind, event.generation.kind) {
        case (.loadCompleted, _), (_, .modelLoaded):
            return RAModelLifecycleChange(
                kind: .loaded,
                modelID: modelID,
                component: event.component,
                event: event
            )
        case (.unloadCompleted, _), (_, .modelUnloaded):
            return RAModelLifecycleChange(
                kind: .unloaded,
                modelID: modelID,
                component: event.component,
                event: event
            )
        default:
            return nil
        }
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-web/packages/core/src/Foundation/EventBus+ModelLifecycle.ts
/**
 * EventBus+ModelLifecycle.ts
 *
 * Typed model-lifecycle stream over the raw SDK event bus — Swift parity:
 * `EventBus+ModelLifecycle.swift` (Public/Extensions/Events).
 *
 * Native commons signals "model loaded/unloaded" on THREE different proto
 * channels depending on the path that performed the work:
 *   1. component-lifecycle events (EVENT_CATEGORY_COMPONENT,
 *      `event.componentLifecycle.currentState`) — the `loadModel` path,
 *   2. model events (`event.model.kind` load/unload completed),
 *   3. LLM generation events (`event.generation.kind` modelLoaded/Unloaded).
 *
 * Which channel fires when is an SDK-internal detail. Consumers should not
 * need to know it — `modelLifecycle()` folds them into one typed stream;
 * `modelLoaded()` / `modelUnloaded()` are pre-filtered views. Swift expresses
 * these as Combine publishers in an `EventBus` extension; TS expresses them
 * as standalone functions over the bus's raw proto stream (AsyncIterable is
 * the Web equivalent of Swift's AsyncStream/AnyPublisher).
 */

import { ComponentLifecycleState, EventCategory } from '@runanywhere/proto-ts/component_types';
import {
  GenerationEventKind,
  ModelEventKind,
  type SDKComponent,
  type SDKEvent as ProtoSDKEvent,
} from '@runanywhere/proto-ts/sdk_events';
import { EventBus } from './EventBus.js';

// MARK: - Typed change

/**
 * One model load/unload transition, decoded from the raw event bus.
 * Swift parity: `RAModelLifecycleChange` (EventBus+ModelLifecycle.swift:25-44).
 */
export interface ModelLifecycleChange {
  /** Whether the model finished loading or was unloaded. */
  kind: 'loaded' | 'unloaded';

  /**
   * Registry id of the affected model. May be empty when the native
   * channel did not carry one (rare; treat as "current model").
   */
  modelId: string;

  /** SDK component slot the change applies to (.llm, .stt, .tts, ...). */
  component: SDKComponent;

  /**
   * The underlying raw event, for consumers that need extra payload
   * fields (progress, framework, error, ...).
   */
  event: ProtoSDKEvent;
}

// MARK: - Decode

/**
 * Decode a raw SDK event into a lifecycle change, or `undefined` when the
 * event is not a load/unload transition. Exposed so consumers can reuse the
 * exact same channel mapping.
 * Swift parity: `EventBus.modelLifecycleChange(from:)`
 * (EventBus+ModelLifecycle.swift:81-128).
 */
export function modelLifecycleChange(event: ProtoSDKEvent): ModelLifecycleChange | undefined {
  // Channel 1: component-lifecycle (the canonical loadModel path).
  if (event.category === EventCategory.EVENT_CATEGORY_COMPONENT) {
    const lifecycle = event.componentLifecycle;
    if (!lifecycle) return undefined;
    switch (lifecycle.currentState) {
      case ComponentLifecycleState.COMPONENT_LIFECYCLE_STATE_READY:
        return {
          kind: 'loaded',
          modelId: lifecycle.modelId,
          component: lifecycle.component,
          event,
        };
      case ComponentLifecycleState.COMPONENT_LIFECYCLE_STATE_NOT_LOADED:
      case ComponentLifecycleState.COMPONENT_LIFECYCLE_STATE_UNLOADING:
      case ComponentLifecycleState.COMPONENT_LIFECYCLE_STATE_SHUTDOWN:
      case ComponentLifecycleState.COMPONENT_LIFECYCLE_STATE_DELETING:
        return {
          kind: 'unloaded',
          modelId: lifecycle.modelId,
          component: lifecycle.component,
          event,
        };
      default:
        return undefined;
    }
  }

  // Channels 2 + 3: model events and LLM generation events.
  const modelId = event.model?.modelId || event.generation?.modelId || '';

  if (
    event.model?.kind === ModelEventKind.MODEL_EVENT_KIND_LOAD_COMPLETED ||
    event.generation?.kind === GenerationEventKind.GENERATION_EVENT_KIND_MODEL_LOADED
  ) {
    return { kind: 'loaded', modelId, component: event.component, event };
  }

  if (
    event.model?.kind === ModelEventKind.MODEL_EVENT_KIND_UNLOAD_COMPLETED ||
    event.generation?.kind === GenerationEventKind.GENERATION_EVENT_KIND_MODEL_UNLOADED
  ) {
    return { kind: 'unloaded', modelId, component: event.component, event };
  }

  return undefined;
}

// MARK: - Typed streams

/**
 * Unified model load/unload stream across all native signal channels.
 * Swift parity: `EventBus.modelLifecycle` (EventBus+ModelLifecycle.swift:58-62).
 *
 * ```ts
 * for await (const change of modelLifecycle()) {
 *   if (change.component === SDKComponent.SDK_COMPONENT_LLM) { ... }
 * }
 * ```
 */
export function modelLifecycle(
  bus: EventBus = EventBus.shared,
): AsyncIterable<ModelLifecycleChange> {
  return lifecycleStream(bus, () => true);
}

/**
 * `modelLifecycle` filtered to load completions.
 * Swift parity: `EventBus.modelLoaded` (EventBus+ModelLifecycle.swift:65-69).
 */
export function modelLoaded(
  bus: EventBus = EventBus.shared,
): AsyncIterable<ModelLifecycleChange> {
  return lifecycleStream(bus, (change) => change.kind === 'loaded');
}

/**
 * `modelLifecycle` filtered to unloads.
 * Swift parity: `EventBus.modelUnloaded` (EventBus+ModelLifecycle.swift:72-76).
 */
export function modelUnloaded(
  bus: EventBus = EventBus.shared,
): AsyncIterable<ModelLifecycleChange> {
  return lifecycleStream(bus, (change) => change.kind === 'unloaded');
}

/**
 * Decode + filter the bus's raw proto stream into lifecycle changes.
 * Mirrors Swift's `events.compactMap(modelLifecycleChange(from:)).filter { }`
 * publisher chain (EventBus+ModelLifecycle.swift:58-76).
 */
function lifecycleStream(
  bus: EventBus,
  predicate: (change: ModelLifecycleChange) => boolean,
): AsyncIterable<ModelLifecycleChange> {
  const source = bus.protoEvents;
  return {
    async *[Symbol.asyncIterator](): AsyncGenerator<ModelLifecycleChange> {
      for await (const event of source) {
        const change = modelLifecycleChange(event);
        if (change && predicate(change)) {
          yield change;
        }
      }
    },
  };
}
